packages/core/src/2d/text/TextUtils.ts

```diff
@@ -0,0 +1,413 @@
+   * @param enableWrapping - whether wrap text to next line when exceeds the width of the container
+   * @param overflowMode - the overflow mode
+   * @param fontStr - the font string
+   * @returns the TextMetrics object
+   */
+  public static measureText(
+    text: string,
+    originWidth: number,
+    originHeight: number,
+    lineSpacing: number,
+    enableWrapping: boolean,
+    overflowMode: OverflowMode,
+    fontStr: string
+  ): TextMetrics {
+    const { _pixelsPerUnit } = TextUtils;
+    const fontSize = TextUtils.measureFont(fontStr);
+    const textContext = TextUtils.textContext();
+    const { context } = textContext;
+    context.font = fontStr;
+    const lines = TextUtils._wordWrap(text, originWidth, enableWrapping, fontStr);
+    const linesLen = lines.length;
+    const lineWidths = new Array<number>();
+    const lineHeight = fontSize + lineSpacing * _pixelsPerUnit;
+    // Calculate max width of all lines.
+    let maxLineWidth = 0;
+    for (let i = 0; i < linesLen; ++i) {
+      const width = Math.ceil(context.measureText(lines[i]).width);
+      if (width > maxLineWidth) {
+        maxLineWidth = width;
+      }
+      lineWidths.push(width);
+    }
+
+    // Reset width and height.
+    const width = Math.min(maxLineWidth, TextUtils._maxWidth);
+    let height = originHeight * _pixelsPerUnit;
+    if (overflowMode === OverflowMode.Overflow) {
+      height = Math.min(lineHeight * linesLen, TextUtils._maxHeight);
+    }
+
+    return {
+      width,
+      height,
+      lines,
+      lineWidths,
+      maxLineWidth,
+      lineHeight,
+      fontSize
+    };
+  }
+
+  /**
+   * Trim canvas.
+   * @returns the width and height after trim, and the image data
+   */
+  public static trimCanvas(): { width: number; height: number; data?: ImageData } {
+    // https://gist.github.com/remy/784508
+
+    const { canvas, context } = TextUtils.textContext();
+    let { width, height } = canvas;
+
+    const imageData = context.getImageData(0, 0, width, height).data;
+    const len = imageData.length;
+
+    let top = -1;
+    let bottom = -1;
+    let left = -1;
+    let right = -1;
+    let data = null;
+    let x;
+    let y;
+
+    for (let i = 0; i < len; i += 4) {
+      if (imageData[i + 3] !== 0) {
+        const idx = i / 4;
+        x = idx % width;
+        y = ~~(idx / width);
+
+        if (top === -1) {
+          top = y;
+        }
+
+        if (left === -1 || x < left) {
+          left = x;
+        }
+
+        if (right === null || x > right) {
+          right = x;
+        }
+
+        if (bottom === null || y > bottom) {
+          bottom = y;
+        }
+      }
+    }
+
+    if (top !== -1) {
+      top = Math.max(0, top - 1);
+      bottom = Math.min(height - 1, bottom + 1);
+      left = Math.max(0, left - 1);
+      right = Math.min(width - 1, right + 1);
+      width = right - left + 1;
+      height = bottom - top + 1;
+      data = context.getImageData(left, top, width, height);
+    }
```

Comment:
Does the return value make sense if the text is not drawn?

Reply:
This has been filtered in TextRenderer